SteamKit2/SteamKit2/Networking/Steam3/TcpConnection.cs

```diff
@@ -157,36 +157,29 @@ private void TryConnect(object sender)
                 return;
             }
 
-            IAsyncResult asyncResult;
-
             try
             {
-                asyncResult = socket!.BeginConnect( CurrentEndPoint, null, null );
+                using var timeoutTokenSource = new CancellationTokenSource( timeout );
+                using var connectCancellation = CancellationTokenSource.CreateLinkedTokenSource( cancellationToken.Token, timeoutTokenSource.Token );
+
+                using ( connectCancellation.Token.Register( s => ( ( Socket )s ).Dispose(), socket ) )
+                {
+                    socket!.Connect( CurrentEndPoint );
+                }
             }
-            catch ( Exception ex )
+            catch ( SocketException ) when ( cancellationToken.IsCancellationRequested )
             {
-                log.LogDebug( nameof( TcpConnection ), "Exception while beginning connection request to {0}: {1}", CurrentEndPoint, ex );
                 ConnectCompleted( false );
                 return;
             }
-
-            if ( WaitHandle.WaitAny( new WaitHandle[] { asyncResult.AsyncWaitHandle, cancellationToken.Token.WaitHandle }, timeout ) == 0 )
-            {
-                try
-                {
-                    socket.EndConnect( asyncResult );
-                    ConnectCompleted( true );
-                }
-                catch ( Exception ex )
-                {
-                    log.LogDebug( nameof( TcpConnection ), "Exception while completing connection request to {0}: {1}", CurrentEndPoint, ex );
-                    ConnectCompleted( false );
-                }
-            }
-            else
+            catch ( Exception ex )
             {
+                log.LogDebug( nameof( TcpConnection ), "Exception while beginning connection request to {0}: {1}", CurrentEndPoint, ex );
                 ConnectCompleted( false );
+                return;
             }
+
+            ConnectCompleted( true );

```

Comment:
```suggestion
            ConnectCompleted( socket.Connected );
```

Would this be saner/more correct?